[stdlib] Generalize some constructs for non-escapable types

lorentey opened the pull request:
This ~quick experimental~ PR generalizes some stdlib constructs to support non-escapable use.

- Generalize `MemoryLayout` to allow querying the layout of nonescapable types
- Allow nonescapable types to conform to `ExpressibleByNilLiteral`
- Generalize `Optional` to support nonescapable wrapped types (itself becoming conditionally escapable)
- Generalize `unsafeBitCast` to allow bit casting from a nonescapable type
- Allow `ManagedBufferPointer.==` to compare managed buffers of noncopyable elements.
- Generalize `withExtendedLifetime` for noncopyable items.
- Add `func extendLifetime(_:)` to avoid a clumsy closure argument with complex lifetimes.
- Allow comparing metatypes of noncopyable and nonescapable types
- Allow generating an `ObjectIdentifier` from noncopyable and nonescapable metatypes
- Start generalizing `Result` for nonescapable `Success` types. Most API is left as is, pending proper lifetime declarations.

rdar://125927256

lorentey committed:
[stdlib] Optional: Allow non-escapable wrapped types

lorentey commented:
@swift-ci test

Azoy commented:
:shipit:

Azoy commented:
Can we get the `ExpressibleByNilLiteral` and `init(_ wrapped: consuming Wrapped)` initializer supported?

lorentey commented:
For `ExpressibleByNilLiteral`, I believe we still need to figure out how to express indefinite lifetime for the result of the `init(nilLiteral:)` initializer.

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey committed:
[stdlib] Generalize MemoryLayout for ~Escapable types

lorentey committed:
[stdlib] ExpressibleByNilLiteral: allow ~Escapable conformers

lorentey committed:
[stdlib] unsafeBitCast: generalize for ~Escapable input types

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey commented:
The last push added generalizations for `MemoryLayout`, `ExpressibleByNilLiteral` and `unsafeBitCast` in addition to a somewhat working `Optional` generalization.

I had to temporarily disable module verification and silence a bogus diagnostic to make this stuff work. Hopefully that'll go away soon.

lorentey committed:
[test] Ritually confess these changes

lorentey commented:
@swift-ci build toolchain

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey committed:
[stdlib] Enable NonescapableTypes

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey commented:
@swift-ci build toolchain

glessard commented:
@swift-ci build toolchain

glessard commented:
@swift-ci please build toolchain

glessard commented:
@swift-ci smoke test

lorentey committed:
[stdlib] Generalize ManagedBufferPointer.==

lorentey committed:
[stdlib] Generalize withExtendedLifetime; add extendLifetime

lorentey committed:
[stdlib] ObjectIdentifier: Allow initialization from metatypes of noncopyable types

lorentey committed:
[stdlib] ObjectIdentifier: Also allow nonescapable types

Azoy commented on stdlib/public/core/ObjectIdentifier.swift:
```diff
@@ -65,6 +65,11 @@ public struct ObjectIdentifier: Sendable {
   }
+
+  @_alwaysEmitIntoClient
+  public init(_ x: any (~Copyable & ~Escapable).Type) {
```
You could just convert the old `Any.Type` initializer and mark it `@_preInverseGenerics`.

lorentey replied:
Oh! I thought `Any.Type` wouldn't ever include noncopyable/nonescapable stuff. I'll put it on the to do list! 👍

lorentey committed:
[stdlib] Start generalizing Result

- Allow nonescapable `Success` types
- Generalize `Sendable` conformance
- Generalize `Result.mapError`
- Generalize `Result.get()`

Do not generalize `Result.init(catching:)` or higher-order functions involving the success value: they cannot possibly work correctly yet.

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey committed:
[stdlib] dependsOn → @lifetime

lorentey committed:
[stdlib] Note some syntactic wrinkles

lorentey committed:
[test][stdlib] Spin tests for Optional generalizations off into a separate file

lorentey committed:
[stdlib] ObjectIdentifier: Work around issue with generalized existentials

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey committed:
[test] Adjust for Optional relying on $NonescapableTypes now

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey committed:
[test] Adjust more test expectations

lorentey committed:
[stdlib] Generalize Optional’s “some” initializer

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey committed:
[stdlib] Generalize Optional.take(), ?? and nil comparisons

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey committed:
[stdlib] Generalize Unsafe[Mutable]BufferPointer.indices

lorentey committed:
[stdlib] Generalize Optional.unsafelyUnwrapped

lorentey committed:
[test] Adjust tests now that Optional supports nonescapable use

`Optional` is now depending on `$NonescapableTypes`, and thus the compiler no longer generates `#else` branches to allow suppressing that feature.

lorentey committed:
[test] api-digester: Add false positives for newly generalized properties

lorentey commented:
Note that there are some known C++ interop failures that I haven't looked at in detail yet. This run is to find out if there is anything else I'm not yet aware of.

benrimmington commented:
Should this PR remove the `unsafeBitCast` function used by C++ interop?

https://github.com/swiftlang/swift/pull/77700#discussion_r1854353785

https://github.com/swiftlang/swift/blob/b6cc118f71acbab0cb076521065c4176d643aecb/stdlib/public/Cxx/CxxSpan.swift#L12-L19

lorentey committed:
[stdlib] Generalize unsafe buffer pointer operations on Slice

`Slice` continues to require `Element` to be copyable, so this is limited to`moveInitializeMemory`, `bindMemory`, `withMemoryRebound` and `assumingMemoryBound`. Generalizing these restores parity with the corresponding operations on unsliced buffer pointers, and reduces the need to artificially rebase the slices only to call these.

lorentey committed:
[stdlib] Roll back generalization of the nil-coalescing operator ??

We cannot currently express its proper lifetime semantics: its result’s lifetime should depend on the intersection of the lifetime of the left argument and the lifetime of the result of the right argument.

`@lifetime(optional, defaultValue.result)` is what we want, but the `.result` part is not currently expressible. (Tying the dependency on the closure argument itself may be a viable option, but we aren’t confident enough to ship it like that yet.)

lorentey committed:
[stdlib] Update syntax for conditional conformances to invertable protocols

lorentey committed:
Revert "[stdlib] unsafeBitCast: generalize for ~Escapable input types"

This reverts commit 0d7cab589f1a7fbcb988b28aec50db768e018539.

lorentey committed:
[test] Update stability expectations

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey committed:
[stdlib] Generalize the signature of type(of:)

`func type(of:)` is wholly magical, but it does have a signature in the stdlib (mostly for documentation purposes), and it currently requires its input to be copyable and escapable.

`type(of:)` is actually usable on all entities, so it seems desirable to update its signature to reflect this.

Additionally, this seems like a good time to mark its exported symbol obsolete. I don’t expect anyone would ever link to it (unless there is/was a bug), so in theory we could also silently remove it — but explicitly marking it as legacy ABI seems the least risky option.

lorentey committed:
[stdlib] Properly generalize ObjectIdentifier.init(Any.Type)

With Slava’s fix in https://github.com/swiftlang/swift/pull/79647, we can now refer to generalized type existentials in ABI stable code.

lorentey committed:
[stdlib] Allow metatypes of noncopyable/nonescapable types to get compared

lorentey committed:
[stdlib] Take out generalization of Optional.unsafelyUnwrapped for now

lorentey commented:
@swift-ci test

lorentey committed:
[test] Temporarily xfail tests broken by lack of is_same_metatype use

lorentey committed:
[AST] Generalize is_same_metatype builtin to take unconstrained metatypes

The builtin’s current is signature is:

```
(Any.Type, Any.Type) -> Bool
```

This needs to be changed to this:

```
(any (~Copyable & ~Escapable).Type, any (~Copyable & ~Escapable).Type) -> Bool
```

This requires a bit of support work in AST synthesis.

rdar://145707064

Co-authored-by: Alejandro Alonso <alejandro_alonso@apple.com>

lorentey committed:
[stdlib] Properly generalize metatype comparisons

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
@swift-ci build toolchain macOS platform

lorentey commented:
@swift-ci build-toolchain linux platform

lorentey commented:
@swift-ci build toolchain linux platform

lorentey committed:
[stdlib] Postpone generalizing Result.flatMapError

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci build toolchain

lorentey committed:
[test] Update list of false positives in source stability test

lorentey commented:
@swift-ci test

lorentey committed:
[stdlib] Mark Unsafe*BufferPointer.indices as safe

lorentey committed:
Reinstate generalization of Optional.unsafelyUnwrapped

lorentey committed:
Add some missing `@lifetime` attributes

We were relying on infererred lifetimes by accident — it’s better to spell things out for now.

lorentey committed:
Generalize Optional._consumingUncheckedUnwrapped()

This is just to round things out — this internal function was sticking out from the public interfaces. No stdlib code needs this generalization yet, but I see no reason to delay it.

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey committed:
[stdlib] Work around ~Escapable breaking `Optional.init` syntax

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci smoke test

lorentey commented:
@swift-ci test source compatibility

lorentey committed:
[stdlib] Adjust lifetime declarations

lorentey commented:
@swift-ci smoke test

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci smoke test

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
```
******************** TEST 'lldb-api :: tools/lldb-server/TestGdbRemoteAttach.py' FAILED ********************
...
failed to attach to process 14830: Operation not permitted
```

Sure

lorentey commented:
It looks like the tests timed out on both macOS and Linux after spending an hour running swiftpm tests with no output.

```
--- bootstrap | 2025-03-21 08:15:31,760 | INFO    | MainThread | helpers | call_output | Line:99 | executing command >>> 'env SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk SWIFTCI_USE_LOCAL_DEPS=1 SWIFTPM_MACOS_DEPLOYMENT_TARGET=12.0 DYLD_LIBRARY_PATH=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/tsc/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/llbuild-macosx-x86_64/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-argument-parser/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-crypto/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-driver/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-system/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-collections/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-asn1/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-certificates/lib SWIFTCI_DISABLE_SDK_DEPENDENT_TESTS=YES SWIFT_EXEC=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc SWIFT_DRIVER_SWIFT_EXEC=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc CC=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/swift-test --parallel --build-path /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64 --configuration release --very-verbose -Xswiftc -module-cache-path -Xswiftc /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/module-cache -Xbuild-tools-swiftc -module-cache-path -Xbuild-tools-swiftc /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/module-cache' with cwd /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swiftpm
Cancelling nested steps due to timeout
Sending interrupt signal to process
--- bootstrap | 2025-03-21 09:18:02,445 | DEBUG   | MainThread | helpers | call_output | Line:108 | ...
```

lorentey commented:
Those tests appear to finish in about a quarter of an hour in successful runs. Hm.

```
# https://ci.swift.org/job/swift-PR-macos/20975/
--- bootstrap | 2025-03-21 10:23:57,831 | INFO    | MainThread | helpers | call_output | Line:99 | executing command >>> 'env SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk SWIFTCI_USE_LOCAL_DEPS=1 SWIFTPM_MACOS_DEPLOYMENT_TARGET=12.0 DYLD_LIBRARY_PATH=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/tsc/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/llbuild-macosx-x86_64/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-argument-parser/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-crypto/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-driver/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-system/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-collections/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-asn1/lib:/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/swift-certificates/lib SWIFTCI_DISABLE_SDK_DEPENDENT_TESTS=YES SWIFT_EXEC=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc SWIFT_DRIVER_SWIFT_EXEC=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc CC=/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/swift-test --parallel --build-path /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64 --configuration release --very-verbose -Xswiftc -module-cache-path -Xswiftc /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/module-cache -Xbuild-tools-swiftc -module-cache-path -Xbuild-tools-swiftc /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/module-cache' with cwd /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swiftpm
--- bootstrap | 2025-03-21 10:40:35,495 | DEBUG   | MainThread | bootstrap | wrapper | Line:64 | Done call to call_swiftpm ...
```

lorentey commented:
The source compat failures are unrelated UXPASSes in Doggie, so that's good news at least

lorentey commented:
@swift-ci smoke test linux

lorentey commented:
@swift-ci test compiler performance

lorentey commented:
@swift-ci clean test

lorentey commented:
@swift-ci benchmark

lorentey committed:
[AST] Temporarily turn off minor assert to avoid source compat regression

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
I've finally managed to locally reproduce the SwiftPM test timeout. `TestMultipleFailureSwiftTestingPackageTests.testFailure1` to `.testFailure9` appear to get hung in [`TypeInfo.fullyQualifiedNameComponents.getter`](https://github.com/pointfreeco/swift-issue-reporting/blob/e876472912e1f4c057001bbabb6ca2c901818841/Sources/IssueReporting/Internal/SwiftTesting.swift#L519-L542) from https://github.com/pointfreeco/swift-issue-reporting.

```
                                          2001 unique name #1 of Z9088c4bc in peer macro @Test expansion #1 of testFailure9 in TestMultipleFailureSwiftTestingTest
s@Sendable ()  (in TestMultipleFailureSwiftTestingPackageTests) + 1  [0x102d92b75]  peer macro @Test expansion #1 of testFailure9 in TestMultipleFailureSwiftTesti
ngTests with unmangled suffix ".swift":3
                                            2001 testFailure9()  (in TestMultipleFailureSwiftTestingPackageTests) + 536  [0x102d93064]  TestMultipleFailureSwiftTe
stingTests.swift:36
                                              2001 __checkFunctionCall<A, B>(_:calling:_:expression:comments:isRequired:sourceLocation:)  (in libTesting.dylib) +
4164  [0x1034559d8]
                                                2001 Issue.record(configuration:)  (in libTesting.dylib) + 816  [0x10347524c]
                                                  2001 static Event.post(_:for:instant:configuration:)  (in libTesting.dylib) + 1096  [0x103425bbc]
                                                    2001 Event._post(in:configuration:)  (in libTesting.dylib) + 492  [0x103425fb4]
                                                      2001 Configuration.handleEvent(_:in:)  (in libTesting.dylib) + 472  [0x103492800]
                                                        2001 partial apply for closure #1 in static Runner._run(_:)  (in libTesting.dylib) + 360  [0x1034b4428]
                                                          2001 partial apply for closure #3 in entryPoint(passing:eventHandler:)  (in libTesting.dylib) + 144  [0x
103406dd8]
                                                            2001 Event.ConsoleOutputRecorder.record(_:in:)  (in libTesting.dylib) + 136  [0x103430448]
                                                              2001 Event.HumanReadableOutputRecorder.record(_:in:verbosity:)  (in libTesting.dylib) + 5052  [0x103
43677c]
                                                                2001 __Expression._expandedDescription(in:)  (in libTesting.dylib) + 168  [0x1034c3a28]
                                                                  2001 TypeInfo.fullyQualifiedName.getter  (in libTesting.dylib) + 52  [0x103485dbc]
                                                                    2001 TypeInfo.fullyQualifiedNameComponents.getter  (in libTesting.dylib) + 120  [0x103485a94]
```

I don't immediately see how this could loop, but it does invoke the `ObjectIdentifier` initializer that I generalized in this PR. It should be straightforward to figure things out from here.

lorentey committed:
[stdlib] Temporarily reinstate original ObjectIdentifier.init as public

The new generalization sometimes causes a runtime hang that I’m still analyzing. The original declaration should be considered more specific for preexisting use cases, eliminating the source compatibility issue.

lorentey commented:
@swift-ci test

lorentey commented:
The last commit will work around the runtime hang in SwiftPM's swift-testing tests, I think. My local build is still in progress, but let's see if CI gets there first.

lorentey commented:
OK, I locally confirmed that the workaround is effective. Let's hope CI agrees, too🤞

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci smoke test

lorentey commented:
@swift-ci build toolchain

lorentey commented:
@swift-ci test source compatibility

lorentey commented:
The debug mode source compatibility suite is expected to fail with an unrelated issue in `ObjectMapper`. The rest of the tests have all passed, so I'm going to hit the big button.
